File: executors/rust/anvil-step/src/outcome.rs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ANLACO
//! What a step gives back.

use crate::{Named, Value};

/// The step met its criterion.
pub const PASS: &str = "pass";
/// The unit does not comply. Information about the DUT.
pub const FAIL: &str = "fail";
/// Could not be judged. Information about the bench or the step.
pub const ERROR: &str = "error";

/// What a step gives back.
///
/// Build it with the constructors below rather than by hand: `status` is a
/// closed vocabulary, and a string Anvil cannot read turns the step into
/// `error` on the other side, with a message naming what you returned
/// (ADR-0019, Rule 2). `"skipped"` is not here on purpose — only the engine
/// produces it.
///
/// **The threshold is not the step's business**: return the measurement and let
/// the engine judge it against the sequence's `limit` (ADR-0008).
#[derive(Debug, Clone, PartialEq)]
pub struct Outcome {
    pub status: &'static str,
    pub message: String,
    pub measured_value: Option<f64>,
    /// Named values the step returns **besides** the measurement. They take no
    /// part in the verdict; `assign` reads them as `result.outputs.<name>`
    /// (ADR-0020 §3).
    pub outputs: Vec<Named>,
}

impl Outcome {
    fn of(status: &'static str, message: impl Into<String>) -> Self {
        Outcome {
            status,
            message: message.into(),
            measured_value: None,
            outputs: Vec::new(),
        }
    }

    /// The step met its criterion.
    pub fn passed(message: impl Into<String>) -> Self {
        Outcome::of(PASS, message)
    }

    /// The unit does not comply. **Information about the DUT** — not about the
    /// bench, and not about your step (ADR-0019, Rule 2).
    pub fn failed(message: impl Into<String>) -> Self {
        Outcome::of(FAIL, message)
    }

    /// The step could not judge: the instrument did not answer, the parameter
    /// made no sense, the bench is not where it should be. Never `failed` for
    /// these — a broken bench says nothing about the unit.
    pub fn error(message: impl Into<String>) -> Self {
        Outcome::of(ERROR, message)
    }

    /// A measurement. The engine judges it against the sequence's `limit`.
    pub fn measured(value: f64) -> Self {
        Outcome {
            status: PASS,
            message: String::new(),
            measured_value: Some(value),
            outputs: Vec::new(),
        }
    }

    /// Free text for the report: what the step did, or why it could not.
    pub fn message(mut self, message: impl Into<String>) -> Self {
        self.message = message.into();
        self
    }

    /// One named output, the ones `assign` reads as `result.outputs.<name>`.
    /// Declare them in the attribute too (`#[step(outputs(name: f64))]`) so
    /// Anvil can check that expression without running the sequence.
    pub fn output(mut self, name: impl Into<String>, value: impl Into<Value>) -> Self {
        self.outputs.push(Named::new(name, value));
        self
    }
}

/// Whatever a step returned, as an `Outcome`.
///
/// A step may return an `Outcome` or, when there is nothing to add:
///
/// - `f64` → a measurement,
/// - `bool` → pass/fail, the simplest step there is,
/// - `()` → passed with no measurement,
/// - `Result<T, E>` → `Err` is **`error`**, never `fail`.
///
/// That last one is how a step reports a bench problem without ceremony, and it
/// matters more here than in Python: a component compiled to WASM aborts on
/// `panic!`, so there is no exception for the SDK to catch. `?` is the way.
pub trait IntoOutcome {
    fn into_outcome(self) -> Outcome;
}

impl IntoOutcome for Outcome {
    fn into_outcome(self) -> Outcome {
        self
    }
}

impl IntoOutcome for f64 {
    fn into_outcome(self) -> Outcome {
        Outcome::measured(self)
    }
}

impl IntoOutcome for bool {
    fn into_outcome(self) -> Outcome {
        if self {
            Outcome::passed("")
        } else {
            Outcome::failed("the step returned false")
        }
    }
}

impl IntoOutcome for () {
    fn into_outcome(self) -> Outcome {
        Outcome::passed("")
    }
}

impl<T: IntoOutcome, E: std::fmt::Display> IntoOutcome for Result<T, E> {
    /// `Err` is **`error`**, never `fail`: a step that blew up says nothing
    /// about the unit under test, and reporting it as a failed unit is the
    /// false red that mirrors ADR-0019's false green.
    fn into_outcome(self) -> Outcome {
        match self {
            Ok(v) => v.into_outcome(),
            Err(e) => Outcome::error(format!("the step could not run: {e}")),
        }
    }
}

File: executors/csharp/src/Anvil.Step/Authoring/Outcome.cs
// SPDX-License-Identifier: Apache-2.0
// Copyright 2026 ANLACO

using System.Collections.ObjectModel;
using System.Globalization;

namespace Anvil.Step;

/// <summary>
/// What a step answers: a verdict, what it measured, and anything else it wants
/// on the report.
/// </summary>
/// <remarks>
/// <para>
/// <b>The threshold is not here.</b> A step measures and the engine judges the
/// value against the <c>limit</c> declared in the sequence (ADR-0008), so the
/// same step serves a board with a different acceptance criterion without being
/// recompiled — and whoever audits reads the criterion in the sequence.
/// </para>
/// <para>
/// The vocabulary is closed: <c>pass</c>, <c>fail</c> and <c>error</c>. A step
/// never answers <c>skipped</c>; only the engine produces that.
/// </para>
/// </remarks>
public sealed class Outcome
{
    /// <summary>The step met its criterion.</summary>
    public const string Pass = "pass";

    /// <summary>The unit does not comply — information about the DUT.</summary>
    public const string Fail = "fail";

    /// <summary>Could not be judged — information about the bench or the step.</summary>
    public const string Error = "error";

    private readonly List<KeyValuePair<string, StepValue>> _outputs;

    private Outcome(string status, string message, double? measured, List<KeyValuePair<string, StepValue>>? outputs)
    {
        Status = status;
        Message = message;
        MeasuredValue = measured;
        _outputs = outputs ?? [];
    }

    /// <summary>The verdict: one of <see cref="Pass"/>, <see cref="Fail"/>, <see cref="Error"/>.</summary>
    public string Status { get; }

    /// <summary>Free text for the report: what the step did, or why it could not.</summary>
    public string Message { get; }

    /// <summary>
    /// The measurement, if the step measures. A <see langword="double"/> and never
    /// a string: how it is written on the wire is the SDK's business, and getting
    /// that wrong loses the measurement silently (ADR-0038 §6).
    /// </summary>
    public double? MeasuredValue { get; }

    /// <summary>
    /// Named values the step returns <b>besides</b> the measurement. They take no
    /// part in the verdict; <c>assign</c> reads them as
    /// <c>result.outputs.&lt;name&gt;</c> (ADR-0020 §3).
    /// </summary>
    public IReadOnlyList<KeyValuePair<string, StepValue>> Outputs =>
        new ReadOnlyCollection<KeyValuePair<string, StepValue>>(_outputs);

    /// <summary>The step met its criterion.</summary>
    /// <param name="message">What it did, for the report.</param>
    public static Outcome Passed(string message = "") => new(Pass, message, null, null);

    /// <summary>
    /// The unit does not comply. <b>Information about the DUT</b>, never about the
    /// bench (ADR-0019, Rule 2).
    /// </summary>
    /// <param name="message">Why the unit does not comply.</param>
    public static Outcome Failed(string message = "") => new(Fail, message, null, null);

    /// <summary>
    /// The step could not judge. <b>Information about the bench, the step or the
    /// configuration</b> — never about the unit (ADR-0019, Rule 2).
    /// </summary>
    /// <param name="message">Why it could not judge.</param>
    public static Outcome Errored(string message = "") => new(Error, message, null, null);

    /// <summary>
    /// The step measured. The engine judges the value against the sequence's
    /// <c>limit</c>; the step does not.
    /// </summary>
    /// <param name="value">What was measured.</param>
    /// <param name="message">What was measured and how, for the report.</param>
    /// <remarks>
    /// A value that is not finite answers <see cref="Errored"/> instead. A NaN
    /// crossing the wire is a measurement nobody can reconstruct afterwards, and
    /// an unreconstructable measurement is a defect even when every test passes
    /// (ADR-0019, Rule 3).
    /// </remarks>
    public static Outcome Measured(double value, string message = "") =>
        double.IsFinite(value)
            ? new Outcome(Pass, message, value, null)
            : Errored(string.Format(
                CultureInfo.InvariantCulture,
                "the step measured {0}, which is not a value a report can carry",
                double.IsNaN(value) ? "NaN" : (value > 0 ? "+infinity" : "-infinity")));

    /// <summary>Adds a named number to the outputs.</summary>
    /// <param name="name">The name <c>assign</c> will read.</param>
    /// <param name="value">The value.</param>
    public Outcome Output(string name, double value) => Add(name, StepValue.Of(value));

    /// <summary>Adds a named text to the outputs.</summary>
    /// <param name="name">The name <c>assign</c> will read.</param>
    /// <param name="value">The value.</param>
    public Outcome Output(string name, string value) => Add(name, StepValue.Of(value));

    /// <summary>Adds a named boolean to the outputs.</summary>
    /// <param name="name">The name <c>assign</c> will read.</param>
    /// <param name="value">The value.</param>
    public Outcome Output(string name, bool value) => Add(name, StepValue.Of(value));

    /// <summary>Adds a named reference to the outputs — how a handle reaches the sequence.</summary>
    /// <param name="name">The name <c>assign</c> will read.</param>
    /// <param name="value">The reference.</param>
    public Outcome Output(string name, Reference value) => Add(name, StepValue.Of(value));

    private Outcome Add(string name, StepValue value)
    {
        ArgumentException.ThrowIfNullOrEmpty(name);
        var copy = new List<KeyValuePair<string, StepValue>>(_outputs)
        {
            new(name, value),
        };
        return new Outcome(Status, Message, MeasuredValue, copy);
    }
}
